rpcclient/errors.go

```diff
@@ -15,163 +16,513 @@ var (
+	"more than one transaction output in a nulldata script": ErrMultiOpReturn,
+
+	// A timelocked transaction.
+	"transaction is not finalized":               ErrNonFinal,
+	"tried to spend coinbase transaction output": ErrNonFinal,
+
+	// A transaction that is locked by BIP68 sequence logic.
+	"transaction's sequence locks on inputs not met": ErrNonBIP68Final,
+
+	// TODO(yy): find/return the following errors in `btcd`.
+	//
+	// A tiny transaction(in non-witness bytes) that is disallowed.
+	// "unmatched btcd error 1": ErrTxTooSmall,
+	// "unmatched btcd error 2": ErrScriptVerifyFlag,
+	// // A transaction with invalid OP codes.
+	// "unmatched btcd error 3": ErrInvalidOpcode,
+	// // Minimally-small transaction(in non-witness bytes) that is
+	// // allowed.
+	// "unmatched btcd error 4": ErrSameNonWitnessData,
+}
+
+// Define an empty bitcoind errors slice, and initialize it with the errors
+// from the above RPCError during package init.
+var BitcoindErrors []error
+
+// MapRPCErr takes an error returned from calling RPC methods from various
+// chain backend and map it to an defined error here. It uses the `BtcdErrMap`
+// defined above, whose keys are btcd error strings and values are errors made
+// from bitcoind error strings.
+//
+// NOTE: we assume neutrino shares the same error strings as btcd.
+func MapRPCErr(rpcErr error) error {
 	// Iterate the map and find the matching error.
-	for keyErr, rejectReason := range RejectReasonMap {
-		// Match the substring.
-		if strings.Contains(btcErr, keyErr) {
-			return rejectReason
+	for btcdErr, err := range BtcdErrMap {
+		// Match it against btcd's error first.
+		if matchErrStr(rpcErr, btcdErr) {
+			return err
+		}
+	}
+
+	// If not found, try to match it against bitcoind's error.
+	for _, err := range BitcoindErrors {
+		if matchErrStr(rpcErr, err.Error()) {
+			return err
 		}
 	}
 
-	// If there's no match, return the error string directly.
-	return btcErr
+	// If not matched, return the original error wrapped.
+	return fmt.Errorf("%w: %v", ErrUndefined, rpcErr)
+}
+
+// matchErrStr takes an error returned from RPC client and matches it against
+// the specified string. If the expected string pattern is found in the error
+// passed, return true. Both the error strings are normalized before matching.
+func matchErrStr(err error, s string) bool {
+	// Replace all dashes found in the error string with spaces.
+	strippedErrStr := strings.Replace(err.Error(), "-", " ", -1)
+
+	// Replace all dashes found in the error string with spaces.
+	strippedMatchStr := strings.Replace(s, "-", " ", -1)
+
+	// Match against the lowercase.
+	return strings.Contains(
+		strings.ToLower(strippedErrStr),
+		strings.ToLower(strippedMatchStr),
+	)
+}
+
+// Init initializes the bitcoind errors slice with the errors from the
+// `BitcoindRPCErr` defined above.
+func init() {
+	BitcoindErrors = make([]error, 0, errSentinel)
+	for i := uint32(0); i < uint32(errSentinel); i++ {
```

Comment:
Can't we just use this loop in `MapRPCErr` above? That way we don't need a package level `init` and also don't have to worry about anyone modifying the global variable.

Reply:
Nice catch! Yeah totally, now switched!